addons/pkg-maven/common/src/main/java/org/commonjava/indy/pkg/maven/content/MetadataMergeListner.java

```diff
@@ -0,0 +1,207 @@
+import org.commonjava.indy.change.event.ArtifactStoreUpdateType;
+import org.commonjava.indy.content.DirectContentAccess;
+import org.commonjava.indy.content.MergedContentAction;
+import org.commonjava.indy.data.StoreDataManager;
+import org.commonjava.indy.model.core.ArtifactStore;
+import org.commonjava.indy.model.core.Group;
+import org.commonjava.indy.model.core.StoreKey;
+import org.commonjava.indy.model.core.StoreType;
+import org.commonjava.indy.pkg.maven.inject.MetadataCache;
+import org.commonjava.indy.subsys.infinispan.CacheHandle;
+import org.commonjava.maven.galley.model.Transfer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This listener will do these tasks:
+ * <ul>
+ *     <li>When there are member changes for a group, delete groups metadata caches to force next regeneration of the metadata files of the group(cascaded)</li>
+ *     <li>When </li>
+ *     <li>When the metadata file changed of a member in a group, delete correspond cache of that file path of the member and group (cascaded)</li>
+ * </ul>
+ */
+@ApplicationScoped
+public class MetadataMergeListner
+        implements MergedContentAction
+{
+    final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    private DirectContentAccess fileManager;
+
+    @Inject
+    private StoreDataManager storeManager;
+
+    @Inject
+    @MetadataCache
+    private CacheHandle<StoreKey, Map> metadataCache;
+
+    /**
+     * Listen to an #{@link ArtifactStorePreUpdateEvent} and clear the metadata cache due to changed memeber in that event
+     *
+     * @param event
+     */
+    public void onStoreUpdate( @Observes final ArtifactStorePreUpdateEvent event )
+    {
+
+        logger.trace( "Got store-update event: {}", event );
+
+        if ( ArtifactStoreUpdateType.UPDATE == event.getType() )
+        {
+            for ( ArtifactStore store : event )
+            {
+                removeMetadataCacheContent( store, event.getChangeMap() );
+            }
+        }
+    }
+
+    private void removeMetadataCacheContent( final ArtifactStore store,
+                                             final Map<ArtifactStore, ArtifactStore> changeMap )
+    {
+        handleStoreDisableOrEnable( store, changeMap );
+
+        handleGroupMembersChanged( store, changeMap );
+    }
+
+    // if a store is disabled/enabled, we should clear its metadata cache and all of its affected groups cache too.
+    private void handleStoreDisableOrEnable(final ArtifactStore store,
+                                            final Map<ArtifactStore, ArtifactStore> changeMap){
+        final ArtifactStore oldStore = changeMap.get( store );
+        if ( store.isDisabled() != oldStore.isDisabled() )
```

Comment:
So this will clear metadata when disabled, then again when re-enabled. Is this what you intended?

Reply:
Yes. Whatever it is enabled/disabled, I  think it is a trigger which will cause the metadata refresh.